File: src/pcap.rs
use std::net::SocketAddr;

pub struct CapturedPacket {
    pub timestamp_us: u64,
    pub src: SocketAddr,
    pub dst: SocketAddr,
    pub payload: Vec<u8>,
}

pub fn write_pcapng(packets: &[CapturedPacket]) -> Vec<u8> {
    let mut buf = Vec::new();
    write_shb(&mut buf);
    write_idb(&mut buf);
    for pkt in packets {
        write_epb(&mut buf, pkt);
    }
    buf
}

fn put_u16(buf: &mut Vec<u8>, v: u16) {
    buf.extend_from_slice(&v.to_le_bytes());
}

fn put_u32(buf: &mut Vec<u8>, v: u32) {
    buf.extend_from_slice(&v.to_le_bytes());
}

fn put_u64(buf: &mut Vec<u8>, v: u64) {
    buf.extend_from_slice(&v.to_le_bytes());
}

fn write_shb(buf: &mut Vec<u8>) {
    let block_len: u32 = 28;
    put_u32(buf, 0x0A0D_0D0A); // Block Type
    put_u32(buf, block_len);
    put_u32(buf, 0x1A2B_3C4D); // Byte-Order Magic
    put_u16(buf, 1); // Major Version
    put_u16(buf, 0); // Minor Version
    put_u64(buf, 0xFFFF_FFFF_FFFF_FFFF); // Section Length (unspecified)
    put_u32(buf, block_len);
}

fn write_idb(buf: &mut Vec<u8>) {
    let block_len: u32 = 20;
    put_u32(buf, 0x0000_0001); // Block Type
    put_u32(buf, block_len);
    put_u16(buf, 101); // LinkType: LINKTYPE_RAW (raw IPv4/IPv6)
    put_u16(buf, 0); // Reserved
    put_u32(buf, 0x0000_FFFF); // SnapLen
    put_u32(buf, block_len);
}

fn write_epb(buf: &mut Vec<u8>, pkt: &CapturedPacket) {
    let frame = build_ipv4_udp_frame(pkt);
    let captured_len = frame.len() as u32;
    let padded_len = (captured_len + 3) & !3;
    let block_len = 32 + padded_len;

    let ts_high = (pkt.timestamp_us >> 32) as u32;
    let ts_low = (pkt.timestamp_us & 0xFFFF_FFFF) as u32;

    put_u32(buf, 0x0000_0006); // Block Type: EPB
    put_u32(buf, block_len);
    put_u32(buf, 0); // Interface ID
    put_u32(buf, ts_high);
    put_u32(buf, ts_low);
    put_u32(buf, captured_len);
    put_u32(buf, captured_len); // Original Packet Length
    buf.extend_from_slice(&frame);

    let padding = padded_len - captured_len;
    for _ in 0..padding {
        buf.push(0);
    }

    put_u32(buf, block_len);
}

fn build_ipv4_udp_frame(pkt: &CapturedPacket) -> Vec<u8> {
    let payload_len = pkt.payload.len();
    let udp_len = 8 + payload_len;
    let ip_total_len = 20 + udp_len;

    let src_ip = match pkt.src {
        SocketAddr::V4(a) => a.ip().octets(),
        _ => [0; 4],
    };
    let dst_ip = match pkt.dst {
        SocketAddr::V4(a) => a.ip().octets(),
        _ => [0; 4],
    };

    let mut frame = Vec::with_capacity(ip_total_len);

    frame.push(0x45); // Version (4) + IHL (5)
    frame.push(0x00); // DSCP + ECN
    frame.extend_from_slice(&(ip_total_len as u16).to_be_bytes());
    frame.extend_from_slice(&[0, 0]); // Identification
    frame.extend_from_slice(&[0x40, 0x00]); // Flags (DF) + Fragment Offset
    frame.push(64); // TTL
    frame.push(17); // Protocol: UDP
    let checksum_offset = frame.len();
    frame.extend_from_slice(&[0, 0]);
    frame.extend_from_slice(&src_ip);
    frame.extend_from_slice(&dst_ip);

    let checksum = ip_checksum(&frame[..20]);
    frame[checksum_offset] = (checksum >> 8) as u8;
    frame[checksum_offset + 1] = (checksum & 0xFF) as u8;

    frame.extend_from_slice(&pkt.src.port().to_be_bytes());
    frame.extend_from_slice(&pkt.dst.port().to_be_bytes());
    frame.extend_from_slice(&(udp_len as u16).to_be_bytes());
    frame.extend_from_slice(&[0, 0]); // UDP checksum (skip)

    frame.extend_from_slice(&pkt.payload);

    frame
}

fn ip_checksum(header: &[u8]) -> u16 {
    let mut sum: u32 = 0;
    for i in (0..header.len()).step_by(2) {
        let word = u16::from_be_bytes([header[i], header[i + 1]]);
        sum += word as u32;
    }
    while sum > 0xFFFF {
        sum = (sum & 0xFFFF) + (sum >> 16);
    }
    !(sum as u16)
}

File: pion/pcap/pcap.go
// Package pcap writes pcapng files compatible with the Rust pcap.rs writer.
// Uses LINKTYPE_RAW (101) with synthetic IPv4/UDP framing.
package pcap

import (
	"encoding/binary"
	"net"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// Packet is a captured UDP packet.
type Packet struct {
	TimestampUS uint64
	Src         net.UDPAddr
	Dst         net.UDPAddr
	Payload     []byte
}

// Recorder collects packets and writes them to a pcapng file.
type Recorder struct {
	mu      sync.Mutex
	packets []Packet
}

// NewRecorder creates a new packet recorder.
func NewRecorder() *Recorder {
	return &Recorder{}
}

// Record adds a packet to the capture buffer.
func (r *Recorder) Record(src, dst net.UDPAddr, payload []byte) {
	buf := make([]byte, len(payload))
	copy(buf, payload)
	r.mu.Lock()
	r.packets = append(r.packets, Packet{
		TimestampUS: uint64(time.Now().UnixMicro()),
		Src:         src,
		Dst:         dst,
		Payload:     buf,
	})
	r.mu.Unlock()
}

// Save writes the accumulated packets to a pcapng file.
func (r *Recorder) Save(filename string) error {
	r.mu.Lock()
	packets := make([]Packet, len(r.packets))
	copy(packets, r.packets)
	r.mu.Unlock()

	if len(packets) == 0 {
		return nil
	}

	if err := os.MkdirAll(filepath.Dir(filename), 0o755); err != nil {
		return err
	}

	data := writePcapng(packets)
	return os.WriteFile(filename, data, 0o644)
}

// Count returns the number of captured packets.
func (r *Recorder) Count() int {
	r.mu.Lock()
	defer r.mu.Unlock()
	return len(r.packets)
}

func writePcapng(packets []Packet) []byte {
	var buf []byte
	buf = writeSHB(buf)
	buf = writeIDB(buf)
	for i := range packets {
		buf = writeEPB(buf, &packets[i])
	}
	return buf
}

func writeSHB(buf []byte) []byte {
	blockLen := uint32(28)
	buf = putU32(buf, 0x0A0D0D0A) // Block Type
	buf = putU32(buf, blockLen)
	buf = putU32(buf, 0x1A2B3C4D)         // Byte-Order Magic
	buf = putU16(buf, 1)                   // Major Version
	buf = putU16(buf, 0)                   // Minor Version
	buf = putU64(buf, 0xFFFFFFFFFFFFFFFF)  // Section Length (unspecified)
	buf = putU32(buf, blockLen)
	return buf
}

func writeIDB(buf []byte) []byte {
	blockLen := uint32(20)
	buf = putU32(buf, 0x00000001) // Block Type
	buf = putU32(buf, blockLen)
	buf = putU16(buf, 101)        // LinkType: LINKTYPE_RAW
	buf = putU16(buf, 0)          // Reserved
	buf = putU32(buf, 0x0000FFFF) // SnapLen
	buf = putU32(buf, blockLen)
	return buf
}

func writeEPB(buf []byte, pkt *Packet) []byte {
	frame := buildIPv4UDPFrame(pkt)
	capturedLen := uint32(len(frame))
	paddedLen := (capturedLen + 3) & ^uint32(3)
	blockLen := 32 + paddedLen

	tsHigh := uint32(pkt.TimestampUS >> 32)
	tsLow := uint32(pkt.TimestampUS & 0xFFFFFFFF)

	buf = putU32(buf, 0x00000006) // Block Type: EPB
	buf = putU32(buf, blockLen)
	buf = putU32(buf, 0) // Interface ID
	buf = putU32(buf, tsHigh)
	buf = putU32(buf, tsLow)
	buf = putU32(buf, capturedLen)
	buf = putU32(buf, capturedLen) // Original Packet Length
	buf = append(buf, frame...)

	padding := paddedLen - capturedLen
	for i := uint32(0); i < padding; i++ {
		buf = append(buf, 0)
	}

	buf = putU32(buf, blockLen)
	return buf
}

func buildIPv4UDPFrame(pkt *Packet) []byte {
	payloadLen := len(pkt.Payload)
	udpLen := 8 + payloadLen
	ipTotalLen := 20 + udpLen

	srcIP := pkt.Src.IP.To4()
	dstIP := pkt.Dst.IP.To4()
	if srcIP == nil {
		srcIP = net.IPv4zero.To4()
	}
	if dstIP == nil {
		dstIP = net.IPv4zero.To4()
	}

	frame := make([]byte, 0, ipTotalLen)

	// IPv4 header
	frame = append(frame, 0x45)                                      // Version + IHL
	frame = append(frame, 0x00)                                      // DSCP + ECN
	frame = append(frame, byte(ipTotalLen>>8), byte(ipTotalLen))     // Total Length
	frame = append(frame, 0, 0)                                      // Identification
	frame = append(frame, 0x40, 0x00)                                // Flags (DF) + Fragment Offset
	frame = append(frame, 64)                                        // TTL
	frame = append(frame, 17)                                        // Protocol: UDP
	checksumOffset := len(frame)
	frame = append(frame, 0, 0)                                      // Header Checksum (placeholder)
	frame = append(frame, srcIP...)
	frame = append(frame, dstIP...)

	// Compute IP checksum
	cs := ipChecksum(frame[:20])
	frame[checksumOffset] = byte(cs >> 8)
	frame[checksumOffset+1] = byte(cs)

	// UDP header
	frame = append(frame, byte(pkt.Src.Port>>8), byte(pkt.Src.Port))
	frame = append(frame, byte(pkt.Dst.Port>>8), byte(pkt.Dst.Port))
	frame = append(frame, byte(udpLen>>8), byte(udpLen))
	frame = append(frame, 0, 0) // UDP checksum (skip)

	// Payload
	frame = append(frame, pkt.Payload...)

	return frame
}

func ipChecksum(header []byte) uint16 {
	var sum uint32
	for i := 0; i+1 < len(header); i += 2 {
		sum += uint32(binary.BigEndian.Uint16(header[i : i+2]))
	}
	for sum > 0xFFFF {
		sum = (sum & 0xFFFF) + (sum >> 16)
	}
	return ^uint16(sum)
}

func putU16(buf []byte, v uint16) []byte {
	b := make([]byte, 2)
	binary.LittleEndian.PutUint16(b, v)
	return append(buf, b...)
}

func putU32(buf []byte, v uint32) []byte {
	b := make([]byte, 4)
	binary.LittleEndian.PutUint32(b, v)
	return append(buf, b...)
}

func putU64(buf []byte, v uint64) []byte {
	b := make([]byte, 8)
	binary.LittleEndian.PutUint64(b, v)
	return append(buf, b...)
}
